ledger/ledger-api-common/src/main/scala/com/digitalasset/platform/akkastreams/dispatcher/SignalDispatcher.scala

```diff
@@ -63,28 +63,38 @@ class SignalDispatcher private () {
   /** Closes this SignalDispatcher by gracefully completing the existing Source subscriptions.
     * For any downstream with pending signals, at least one such signal will be sent first.
     */
-  def shutdown(): Future[Unit] = shutdownInternal(_.complete())
+  def shutdown(): Future[Unit] =
+    shutdownInternal { source =>
+      source.complete()
+      source.watchCompletion()
+    }
 
   /** Closes this SignalDispatcher by failing the existing Source subscriptions with the provided throwable. */
-  def fail(throwable: Throwable): Future[Unit] =
-    shutdownInternal(_.fail(throwable)).transform {
-      // This throwable is expected so map to Success
-      case Failure(`throwable`) => Success(())
-      case other => other
-    }(ExecutionContext.parasitic)
+  def fail(newThrowable: () => Throwable): Future[Unit] =
+    shutdownInternal { source =>
+      implicit val ec: ExecutionContext = scala.concurrent.ExecutionContext.parasitic
+      val throwable = newThrowable()
+      source.fail(throwable)
+      source
+        .watchCompletion()
+        .transform {
+          // This throwable is expected so map to Success
+          case Failure(`throwable`) =>
+            Success(())
+          case other =>
```

Comment:
If`Failure` is expected, do we actually expect `other` here or we do not actually care?

Reply:
you're right. Only the specific `throwable` instance is expected. I can just use recoverWith since it take a `PartialFunction`